File: find-lucky-integer-in-an-array.ts
{
    function findLucky(arr: number[]): number {
        let result = -1;
        const hash = new Map();

        for(let i = 0; i < arr.length; i++) {
            hash.set(arr[i], (hash.get(arr[i]) || 0) + 1)
        }

        hash.forEach((num, key) => {
            if (num === key)
                result = Math.max(result, key)
        })
        
        return result
    };

    console.log(findLucky([2,2,3,4]))
}

File: find-lucky-integer-in-an-array.rs
use std::collections::HashMap;

fn find_lucky(arr: Vec<i32>) -> i32 {
    let mut freq = HashMap::new();
    let mut result = -1;

    for &num in &arr {
        *freq.entry(num).or_insert(0) += 1;
    }

    for (&key, &value) in &freq {
        if key == value {
            result = result.max(key);
        }
    }

    result
}
